### PR TITLE
Feature: add subcommand for merging binaries

### LINKED ISSUE #79: QEMU support
Supporting QEMU (that is, the Espressif fork) really boils down to:
- Ability to convert an ELF to BIN (see the other issue - #78 )
- Ability to merge `bootloader.bin`, `partitions.bin` and the application BIN into a single BIN, as done [here](https://github.com/ivmarkov/rust-esp32-std-hello/blob/main/qemu.sh#L10)

Given that `bootloader.bin` and *the* single-factory-app version of `partitions.bin` that we need for QEMU are already "embedded" in `espflash`, that extension might be quite a minor one...

### PR DESCRIPTION
**Pull request by JurajSadel**
Feature: add the option for merging bootloader, partition-table and application binary into a single binary.
Closes #79 

Note: To get a chip revision, the connection is needed, for simplicity, the revision None is used.

### CODE REVIEW

File: espflash/src/cli/mod.rs [hunk 1]
```diff
@@ -121,6 +124,65 @@ pub fn save_elf_as_image(
         }
     }
 
+    // merge_bin is TRUE
+    // merge bootloader, partition table and app binaries
+    // basic functionality, only merge 3 binaries
+    if merge {
+        use std::io::Write;
```
Comment (bjoernQ) [inline espflash/src/cli/mod.rs:131]: not sure we use this import style anywhere in this repo? maybe just move it to the top?

File: espflash/src/cli/mod.rs [hunk 2]
```diff
@@ -121,6 +124,65 @@ pub fn save_elf_as_image(
...
+        }
+
+        let mut file = fs::OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(merged_bin)
+            .unwrap();
+
+        for segment in image.flash_segments() {
+            let buf =
```
Comment (bjoernQ) [inline espflash/src/cli/mod.rs:178]: can we find a better name? like `padding_bytes` or similar?

File: espflash/src/cli/mod.rs [hunk 3]
```diff
@@ -121,6 +124,65 @@ pub fn save_elf_as_image(
...
+                vec![b'\xFF'; segment.addr as usize - file.metadata().unwrap().len() as usize];
```
Comment (bjoernQ) [inline espflash/src/cli/mod.rs:179]: I guess `0xffu8` would be easier to read than `b'\xFF'`

**Comment by bjoernQ** [comment]
besides some nitpicking it looks good to me - hopefully we don't get merge conflicts with the rather large draft PR addressing support for ESP32C2

File: espflash/src/cli/mod.rs [hunk 4]
```diff
@@ -121,6 +124,65 @@ pub fn save_elf_as_image(
...
+        } else {
+            None
+        };
+
+        // To get a chip revision, the connection is needed
+        // For simplicity, the revision None is used
+        let image =
+            chip.get_flash_image(&image, bootloader, partition_table, image_format, None)?;
+
+        let merged_bin = format!("merged_{path}", path = &path.to_str().unwrap());
```
Comment (MabezDev) [inline espflash/src/cli/mod.rs:165]: And this unwrap

File: espflash/src/cli/mod.rs [hunk 5]
```diff
@@ -121,6 +124,65 @@ pub fn save_elf_as_image(
...
+
+        if Path::new(&merged_bin).exists() {
+            fs::remove_file(&merged_bin).unwrap();
```
Comment (MabezDev) [inline espflash/src/cli/mod.rs:168]: Ditto

File: espflash/src/cli/mod.rs [hunk 6]
```diff
@@ -121,6 +124,65 @@ pub fn save_elf_as_image(
...
+        }
+
+        let mut file = fs::OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(merged_bin)
+            .unwrap();
```
Comment (MabezDev) [inline espflash/src/cli/mod.rs:175]: Ditto

**Review by MabezDev — CHANGES_REQUESTED** [review]

**Comment by JurajSadel** [comment]
Thank you @bjoernQ, @MabezDev for the reviews!

**Review by MabezDev — APPROVED** [review]
LGTM! Haven't actually tested with QEMU, but it looks like it should work fine :)

**Pull request merged**